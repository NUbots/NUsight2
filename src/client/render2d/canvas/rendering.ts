import { Transform } from '../../math/transform'
import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'
import { LineAppearance } from '../appearance/line_appearance'
import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { MarkerGeometry } from '../geometry/marker_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'
import { Group } from '../object/group'
import { Object2d } from '../object/object2d'
import { Shape } from '../object/shape'

import { renderArc } from './arc'
import { renderArrow } from './arrow'
import { renderCircle } from './circle'
import { renderLine } from './line'
import { renderMarker } from './marker'
import { renderPath } from './path'
import { renderPolygon } from './polygon'
import { renderText } from './text'

export function renderObject2d(ctx: CanvasRenderingContext2D, obj: Object2d, world: Transform) {

  if (obj instanceof Group) {
    const objWorld = world.clone().then(obj.transform)

    for (const o of obj.children) {
      ctx.save()
      applyTransform(ctx, obj.transform)
      renderObject2d(ctx, o, objWorld)
      ctx.restore()
    }
  } else if (obj instanceof Shape) {
    if (obj.geometry instanceof ArcGeometry) {
      renderArc(ctx, obj)
    } else if (obj.geometry instanceof ArrowGeometry) {
      renderArrow(ctx, obj)
    } else if (obj.geometry instanceof CircleGeometry) {
      renderCircle(ctx, obj)
    } else if (obj.geometry instanceof LineGeometry) {
      renderLine(ctx, obj)
    } else if (obj.geometry instanceof MarkerGeometry) {
      renderMarker(ctx, obj)
    } else if (obj.geometry instanceof PathGeometry) {
      renderPath(ctx, obj)
    } else if (obj.geometry instanceof PolygonGeometry) {
      renderPolygon(ctx, obj)
    } else if (obj.geometry instanceof TextGeometry) {
      renderText(ctx, obj, world)
    } else {
      throw new Error(`Unsupported geometry type: ${obj.geometry}`)
    }
  } else {
    throw new Error(`Unsupported Object2d type: ${obj}`)
  }
}

export function applyTransform(ctx: CanvasRenderingContext2D, transform: Transform): void {
  ctx.translate(transform.translate.x, transform.translate.y)
  ctx.scale(transform.scale.x, transform.scale.y)
  ctx.rotate(transform.rotate * (transform.anticlockwise ? 1 : -1))
}

export const hexToRGB = (hex: string): { r: number, g: number, b: number} => {
  const result = /^#([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})/.exec(hex)

  if (result === null) {
    throw Error(`Color ${hex} was not a hex color`)
  } else {
    return {
      r: parseInt(result[1], 16),
      g: parseInt(result[2], 16),
      b: parseInt(result[3], 16),
    }
  }
}

export function applyAppearance(ctx: CanvasRenderingContext2D, appearance: Appearance): void {

  if (appearance instanceof BasicAppearance) {

    const fill = hexToRGB(appearance.fill.color)
    const fA = appearance.fill.alpha
    ctx.fillStyle = `rgba(${fill.r}, ${fill.g}, ${fill.b}, ${fA})`

    const stroke = hexToRGB(appearance.stroke.color)
    const sA = appearance.stroke.alpha
    ctx.lineWidth = appearance.stroke.width
    ctx.strokeStyle = `rgba(${stroke.r}, ${stroke.g}, ${stroke.b}, ${sA})`
  } else if (appearance instanceof LineAppearance) {
    ctx.lineCap = appearance.stroke.cap
    ctx.lineDashOffset = appearance.stroke.dashOffset
    ctx.lineJoin = appearance.stroke.join


    const stroke = hexToRGB(appearance.stroke.color)
    const sA = appearance.stroke.alpha
    ctx.lineWidth = appearance.stroke.width
    ctx.strokeStyle = `rgba(${stroke.r}, ${stroke.g}, ${stroke.b}, ${sA})`
  } else {
    throw new Error(`Unsupported appearance type: ${appearance}`)
  }
}
