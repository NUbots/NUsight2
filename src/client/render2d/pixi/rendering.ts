import { createTransformer } from 'mobx-utils'
import { Graphics } from 'pixi.js'
import { Container } from 'pixi.js'
import { DisplayObject } from 'pixi.js'

import { Transform } from '../../math/transform'
import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'
import { LineAppearance } from '../appearance/line_appearance'
import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { MarkerGeometry } from '../geometry/marker_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'
import { Group } from '../object/group'
import { Object2d } from '../object/object2d'
import { Shape } from '../object/shape'

import { renderArc } from './arc'
import { renderArrow } from './arrow'
import { renderCircle } from './circle'
import { renderLine } from './line'
import { renderMarker } from './marker'
import { renderPath } from './path'
import { renderPolygon } from './polygon'
import { renderText } from './text'

export const renderObject2d = createTransformer((obj: Object2d): DisplayObject => {

  if (obj instanceof Group) {
    const g = new Container()

    g.scale.x = obj.transform.scale.x
    g.scale.y = obj.transform.scale.x
    g.x = obj.transform.translate.x
    g.y = obj.transform.translate.y
    g.rotation = obj.transform.rotate

    obj.children.forEach(o => {
      g.addChild(renderObject2d(o))
    })

    return g

  } else if (obj instanceof Shape) {
    if (obj.geometry instanceof ArcGeometry) {
      return renderArc(obj)
    } else if (obj.geometry instanceof ArrowGeometry) {
      return renderArrow(obj)
    } else if (obj.geometry instanceof CircleGeometry) {
      return renderCircle(obj)
    } else if (obj.geometry instanceof LineGeometry) {
      return renderLine(obj)
    } else if (obj.geometry instanceof MarkerGeometry) {
      return renderMarker(obj)
    } else if (obj.geometry instanceof PathGeometry) {
      return renderPath(obj)
    } else if (obj.geometry instanceof PolygonGeometry) {
      return renderPolygon(obj)
    } else if (obj.geometry instanceof TextGeometry) {
      return renderText(obj)
    } else {
      throw new Error(`Unsupported geometry type: ${obj.geometry}`)
    }
  } else {
    throw new Error(`Unsupported Object2d type: ${obj}`)
  }
})

const toPixiColor = (style: string): [number, number] => {
  if (style === 'transparent') {
    return [0, 0]
  }
  let result = /^#([A-Fa-f0-9])([A-Fa-f0-9])([A-Fa-f0-9])$/.exec(style)
  if (result !== null) {
    return[parseInt(result[1] + result[1] + result[2] + result[2] + result[3] + result[3], 16), 1]
  }
  result = /^#([A-Fa-f0-9]{6})$/.exec(style)
  if (result !== null) {
    return [parseInt(result[1], 16), 1]
  }

  throw new Error('Pixi cannot handle non hex colours')
}

export function applyAppearance(obj: Graphics, appearance: Appearance, exec: (obj: Graphics) => void): void {

  if (appearance instanceof BasicAppearance) {
    const line = toPixiColor(appearance.strokeStyle)
    const fill = toPixiColor(appearance.fillStyle)
    obj.lineStyle(appearance.lineWidth, line[0], line[1])
    obj.beginFill(fill[0], fill[1])
    exec(obj)
    obj.endFill()
  } else if (appearance instanceof LineAppearance) {
    const line = toPixiColor(appearance.strokeStyle)
    obj.lineStyle(appearance.lineWidth, line[0], line[1])
    exec(obj)
    // ctx.lineCap = appearance.lineCap
    // ctx.lineDashOffset = appearance.lineDashOffset
    // ctx.lineJoin = appearance.lineJoin
    // ctx.lineWidth = appearance.lineWidth
    // ctx.strokeStyle = appearance.strokeStyle
  } else {
    throw new Error(`Unsupported appearance type: ${appearance}`)
  }
}
