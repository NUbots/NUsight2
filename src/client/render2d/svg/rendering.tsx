import { observer } from 'mobx-react'
import * as React from 'react'

import { Transform } from '../../math/transform'
import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'
import { LineAppearance } from '../appearance/line_appearance'
import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { MarkerGeometry } from '../geometry/marker_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'
import { Group as GroupGeometry } from '../object/group'
import { Object2d } from '../object/object2d'
import { Shape } from '../object/shape'

import { Arc } from './arc'
import { Arrow } from './arrow'
import { Circle } from './circle'
import { Group } from './group'
import { Line } from './line'
import { Marker } from './marker'
import { Path } from './path'
import { Polygon } from './polygon'
import { Text } from './text'

export function toSvgProps(appearance: Appearance) {
  if (appearance instanceof BasicAppearance) {
    return {
      fill: appearance.fillStyle,
      strokeWidth: appearance.lineWidth,
      stroke: appearance.strokeStyle,
    }
  } else if (appearance instanceof LineAppearance) {
    return {
      strokeLinecap: appearance.lineCap,
      strokeLinejoin: appearance.lineJoin,
      strokeDashoffset: appearance.lineDashOffset,
      strokeWidth: appearance.lineWidth,
      stroke: appearance.strokeStyle,
      vectorEffect: appearance.nonScalingStroke ? 'non-scaling-stroke' : '',
    }
  } else {
    throw new Error(`Unsupported appearance type ${appearance}`)
  }
}

export function toSvgTransform(transform: Transform): string {
  const s = transform.scale
  const r = (180.0 / Math.PI) * transform.rotate // SVG rotations are in degrees
  const t = transform.translate
  return `translate(${t.x}, ${t.y}) rotate(${r}) scale(${s.x}, ${s.y})`
}

type Props = { obj: Object2d, world: Transform }

export const GeometryView = observer(({ obj, world }: Props): JSX.Element => {
  if (obj instanceof GroupGeometry) {
    return <Group model={obj} world={world}/>
  } else if (obj instanceof Shape) {
    if (obj.geometry instanceof ArcGeometry) {
      return <Arc model={obj} world={world}/>
    } else if (obj.geometry instanceof ArrowGeometry) {
      return <Arrow model={obj} world={world}/>
    } else if (obj.geometry instanceof CircleGeometry) {
      return <Circle model={obj} world={world}/>
    } else if (obj.geometry instanceof LineGeometry) {
      return <Line model={obj} world={world}/>
    } else if (obj.geometry instanceof MarkerGeometry) {
      return <Marker model={obj} world={world}/>
    } else if (obj.geometry instanceof PathGeometry) {
      return <Path model={obj} world={world}/>
    } else if (obj.geometry instanceof PolygonGeometry) {
      return <Polygon model={obj} world={world}/>
    } else if (obj.geometry instanceof TextGeometry) {
      return <Text model={obj} world={world}/>
    } else {
      throw new Error(`Unsupported geometry type: ${obj}`)
    }
  } else {
    throw new Error(`Unsupported geometry type: ${obj}`)
  }
})
