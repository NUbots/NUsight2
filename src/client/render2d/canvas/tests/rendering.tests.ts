import { hexToRGB } from '../rendering'


describe('hexToRGB', () => {

  it('converts hex colours to rgb', () => {

    expect(hexToRGB('#AABBCC')).toEqual({
      r: 0xAA,
      g: 0xBB,
      b: 0xCC,
    })

  })

  it('throws if not a hex colour', () => {

    expect(() => hexToRGB('#FFF')).toThrowErrorMatchingSnapshot()
    expect(() => hexToRGB('red')).toThrowErrorMatchingSnapshot()

  })
})
