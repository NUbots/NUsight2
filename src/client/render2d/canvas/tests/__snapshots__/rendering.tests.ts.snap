// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hexToRGB throws if not a hex colour 1`] = `"Color #FFF is not a hex color"`;

exports[`hexToRGB throws if not a hex colour 2`] = `"Color red is not a hex color"`;
